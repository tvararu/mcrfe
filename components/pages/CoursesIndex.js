import gql from "graphql-tag";
import Link from "next/link";
import Query from "../elements/Query";
import GovukMain from "../govuk/Main";
import GovukBreadcrumbs from "../govuk/Breadcrumbs";
import CourseStatusTag from '../elements/CourseStatusTag'

const allProvidersQuery = gql`
  query allProviders($providerCode: String!) {
    allProviders(condition: { providerCode: $providerCode }) {
      nodes {
        providerName
        coursesByProviderId(orderBy: COURSE_CODE_ASC) {
          nodes {
            courseCode
            providerByAccreditingProviderId {
              providerName
            }
            name
            courseSitesByCourseId {
              nodes {
                publish
                status
                applicationsAcceptedFrom
                vacStatus
              }
            }
          }
        }
      }
    }
    allCourseEnrichments(
      condition: { providerCode: $providerCode }
      orderBy: CREATED_AT_DESC
    ) {
      nodes {
        lastPublishedTimestampUtc
        status

        ucasCourseCode
      }
    }
  }
`;

const TableHeader = () => (
  <thead className="govuk-table__head">
    <style jsx>{`
      abbr {
        text-decoration: underline dotted;
      }

      .w-33 {
        width: 33%;
      }
    `}</style>
    <tr className="govuk-table__row">
      <th className="govuk-table__header w-33">Course</th>
      <th className="govuk-table__header">UCAS Status</th>
      <th className="govuk-table__header">Content</th>
      <th className="govuk-table__header">
        Is it on <abbr title="Find postgraduate teacher training">Find</abbr>?
      </th>
      <th className="govuk-table__header">Applications</th>
      <th className="govuk-table__header">Vacancies</th>
    </tr>
  </thead>
);

const siteIsFindable = ({ publish, status }) =>
  publish === "Y" && status === "R";

const courseIsFindable = course =>
  course.courseSitesByCourseId.nodes.some(siteIsFindable);

const courseIsNew = course =>
  course.courseSitesByCourseId.nodes.some(({ status }) => status === "N");

const courseUcasStatus = course => {
  if (courseIsFindable(course)) return "Running";
  if (courseIsNew(course)) return "New – not yet running";

  return "Not running";
};

const courseIsRunning = course => courseUcasStatus(course) === "Running";

const siteOpenForApplications = site =>
  site.applicationsAcceptedFrom &&
  new Date(site.applicationsAcceptedFrom) < new Date();

const siteHasVacancies = site => site.vacStatus !== "";

const courseOpenForApplications = course =>
  course.courseSitesByCourseId.nodes
    .filter(siteIsFindable)
    .filter(siteOpenForApplications)
    .some(siteHasVacancies);

const courseHasVacancies = course =>
  course.courseSitesByCourseId.nodes
    .filter(siteIsFindable)
    .some(siteHasVacancies);

const TableRow = ({ providerCode, course, latestEnrichment }) => (
  <tr className="govuk-table__row">
    <td className="govuk-table__cell">
      <Link
        as={`/organisations/${providerCode}/courses/${course.courseCode}`}
        href={`/organisations/courses/show?providerCode=${providerCode}&courseCode=${
          course.courseCode
        }`}
        prefetch
      >
        <a className="govuk-link govuk-heading-s govuk-!-margin-bottom-0">
          {course.name} ({course.courseCode})
        </a>
      </Link>
      <span className="govuk-body-s">QTS full time</span>
    </td>
    <td className="govuk-table__cell">{courseUcasStatus(course)}</td>
    <td className="govuk-table__cell">
      {courseIsRunning(course) && (
        <CourseStatusTag enrichment={latestEnrichment} />
      )}
    </td>
    <td className="govuk-table__cell">
      {courseIsFindable(course) ? (
        <a
          className="govuk-link"
          href={`https://bat-dev-search-and-compare-ui-app.azurewebsites.net/course/${providerCode}/${
            course.courseCode
          }`}
        >
          Yes - view online
        </a>
      ) : (
        "No"
      )}
    </td>
    <td className="govuk-table__cell">
      {courseIsRunning(course)
        ? courseOpenForApplications(course)
          ? "Open"
          : "Closed"
        : ""}
    </td>
    <td className="govuk-table__cell">
      {courseIsRunning(course) ? (
        <>
          {courseHasVacancies(course) ? "Yes" : "No"} (
          <a
            className="govuk-link"
            href={`/organisations/${providerCode}/courses/${
              course.courseCode
            }/vacancies`}
          >
            Edit
          </a>
          )
        </>
      ) : (
        ""
      )}
    </td>
  </tr>
);

const sortByNameAndCourseCode = (c1, c2) => {
  const c1str = c1.name + c1.courseCode;
  const c2str = c2.name + c2.courseCode;
  return c1str < c2str ? -1 : c1str > c2str ? 1 : 0;
};

const Table = ({ providerCode, courses, enrichments }) => (
  <table className="govuk-table">
    <TableHeader />
    <tbody className="govuk-table__body">
      {courses.sort(sortByNameAndCourseCode).map(course => (
        <TableRow
          key={course.courseCode}
          providerCode={providerCode}
          course={course}
          latestEnrichment={enrichments.find(
            ({ ucasCourseCode }) => ucasCourseCode === course.courseCode
          )}
        />
      ))}
    </tbody>
  </table>
);

const AddNewCourse = ({ top = false }) => (
  <>
    <a
      className="govuk-button govuk-!-margin-bottom-2"
      rel="noopener noreferrer"
      target="_blank"
      href="https://docs.google.com/forms/d/e/1FAIpQLSeatJO2ZuqM-fnRJxEo6IIF0hZIU63JGnx0sDXO6Ulax7U_bA/viewform?usp=pp_url&amp;entry.1033530353=tim.abell%2B4%40digital.education.gov.uk&amp;entry.158771972=2C4"
    >
      Add a new course
    </a>
    <p className={`govuk-body-s${top ? " govuk-!-margin-bottom-6" : ""} `}>
      You’ll be taken to a Google Form to fill in your course details.
    </p>
  </>
);

const Content = ({ providerCode, providerName, courses, enrichments }) => {
  const coursesByAccreditingProvider = courses.reduce((cs, course) => {
    const accreditingProvider = course.providerByAccreditingProviderId;
    const key = accreditingProvider
      ? accreditingProvider.providerName
      : providerName;
    cs[key] = (cs[key] || []).concat(course);
    return cs;
  }, {});

  const selfAccreditedCourses = coursesByAccreditingProvider[providerName];
  delete coursesByAccreditingProvider[providerName];

  return (
    <>
      <p className="govuk-body">Use this section to:</p>

      <ul className="govuk-list govuk-list--bullet govuk-!-margin-bottom-7">
        <li>write about each course</li>
        <li>preview and publish courses</li>
        <li>copy content between courses</li>
      </ul>

      {courses.length > 10 && <AddNewCourse top />}

      <section>
        {selfAccreditedCourses && (
          <Table
            providerCode={providerCode}
            courses={selfAccreditedCourses}
            enrichments={enrichments}
          />
        )}

        {Object.keys(coursesByAccreditingProvider)
          .sort()
          .map(providerName => (
            <React.Fragment key={providerName}>
              <h2 className="govuk-heading-m">
                <span className="govuk-caption-m">Accredited body</span>
                {providerName}
              </h2>

              <Table
                providerCode={providerCode}
                courses={coursesByAccreditingProvider[providerName]}
                enrichments={enrichments}
              />
            </React.Fragment>
          ))}
      </section>

      <AddNewCourse />
    </>
  );
};

export default ({ providerCode }) => (
  <Query query={allProvidersQuery} variables={{ providerCode }}>
    {({ data }) => {
      const provider = data.allProviders.nodes[0];
      const courses = provider.coursesByProviderId.nodes;
      const enrichments = data.allCourseEnrichments.nodes;
      return (
        <>
          <GovukBreadcrumbs
            crumbs={[
              { href: "/", text: "Organisations" },
              {
                as: `/organisations/${providerCode}`,
                href: `/organisations?providerCode=${providerCode}`,
                text: provider.providerName
              }
            ]}
          />
          <GovukMain>
            <h1 className="govuk-heading-xl">Courses</h1>
            <Content
              providerName={provider.providerName}
              providerCode={providerCode}
              courses={courses}
              enrichments={enrichments}
            />
          </GovukMain>
        </>
      );
    }}
  </Query>
);
