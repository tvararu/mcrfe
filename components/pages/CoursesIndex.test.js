import renderer from "react-test-renderer";
import CoursesIndex from "./CoursesIndex";

jest.mock("../elements/Query", () => ({ children }) =>
  children({
    data: {
      allProviders: {
        nodes: [
          {
            providerName: "Acme",
            coursesByProviderId: {
              nodes: [
                {
                  courseCode: "X100",
                  providerByAccreditingProviderId: null,
                  name: "Primary",
                  courseSitesByCourseId: {
                    nodes: [
                      {
                        publish: "Y",
                        status: "R",
                        applicationsAcceptedFrom: "2018-10-09",
                        vacStatus: "F"
                      }
                    ]
                  }
                },
                {
                  courseCode: "X101",
                  providerByAccreditingProviderId: {
                    providerName: "Acme Accrediting"
                  },
                  name: "Mathematics",
                  courseSitesByCourseId: {
                    nodes: [
                      {
                        publish: "N",
                        status: "R",
                        applicationsAcceptedFrom: "2018-10-09",
                        vacStatus: "F"
                      }
                    ]
                  }
                },
                {
                  courseCode: "X102",
                  providerByAccreditingProviderId: {
                    providerName: "Acme Other"
                  },
                  name: "Physics",
                  courseSitesByCourseId: {
                    nodes: [
                      {
                        publish: "N",
                        status: "N",
                        applicationsAcceptedFrom: "2018-10-09",
                        vacStatus: "F"
                      }
                    ]
                  }
                },
                ...Array(10)
                  .fill(null)
                  .map((_, index) => ({
                    courseCode: `Y${index + 1}`,
                    providerByAccreditingProviderId: {
                      providerName: "Acme Lots"
                    },
                    name: "English",
                    courseSitesByCourseId: {
                      nodes: []
                    }
                  })),
                {
                  courseCode: `Y0`,
                  providerByAccreditingProviderId: {
                    providerName: "Acme Lots"
                  },
                  name: "English",
                  courseSitesByCourseId: {
                    nodes: []
                  }
                }
              ]
            }
          }
        ]
      },
      allCourseEnrichments: {
        nodes: [
          {
            lastPublishedTimestampUtc: "2018-09-10T13:52:19.665409",
            status: 1,
            ucasCourseCode: "X100"
          },
          {
            lastPublishedTimestampUtc: "2018-09-10T12:41:13.796804",
            status: 1,
            ucasCourseCode: "X100"
          }
        ]
      }
    }
  })
);

describe("CoursesIndex", () => {
  it("renders correctly", () => {
    const component = renderer.create(<CoursesIndex providerCode="A01" />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
