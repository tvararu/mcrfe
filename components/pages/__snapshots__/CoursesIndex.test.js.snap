// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesIndex renders correctly 1`] = `
Array [
  <nav
    className="govuk-breadcrumbs"
  >
    <ol
      className="govuk-breadcrumbs__list"
    >
      <li
        className="govuk-breadcrumbs__list-item"
      >
        <a
          className="govuk-breadcrumbs__link"
        >
          Organisations
        </a>
      </li>
      <li
        className="govuk-breadcrumbs__list-item"
      >
        <a
          className="govuk-breadcrumbs__link"
        >
          Acme
        </a>
      </li>
    </ol>
  </nav>,
  <main
    className="govuk-main-wrapper"
    id="main-content"
    role="main"
  >
    <h1
      className="govuk-heading-xl"
    >
      Courses
    </h1>
    <p
      className="govuk-body"
    >
      Use this section to:
    </p>
    <ul
      className="govuk-list govuk-list--bullet govuk-!-margin-bottom-7"
    >
      <li>
        write about each course
      </li>
      <li>
        preview and publish courses
      </li>
      <li>
        copy content between courses
      </li>
    </ul>
    <section>
      <table
        className="govuk-table"
      >
        <thead
          className="jsx-621630794 govuk-table__head"
        >
          <tr
            className="jsx-621630794 govuk-table__row"
          >
            <th
              className="jsx-621630794 govuk-table__header w-33"
            >
              Course
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              UCAS Status
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Content
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Is it on 
              <abbr
                className="jsx-621630794"
                title="Find postgraduate teacher training"
              >
                Find
              </abbr>
              ?
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Applications
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Vacancies
            </th>
          </tr>
        </thead>
        <tbody
          className="govuk-table__body"
        >
          <tr
            className="govuk-table__row"
          >
            <td
              className="govuk-table__cell"
            >
              <a
                className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
              >
                Primary
                 (
                X100
                )
              </a>
              <span
                className="govuk-body-s"
              >
                QTS full time
              </span>
            </td>
            <td
              className="govuk-table__cell"
            >
              Running
            </td>
            <td
              className="govuk-table__cell"
            >
              <div
                className="jsx-268204376 govuk-tag phase-tag--small phase-tag--published"
              >
                Published
              </div>
            </td>
            <td
              className="govuk-table__cell"
            >
              <a
                className="govuk-link"
                href="https://bat-dev-search-and-compare-ui-app.azurewebsites.net/course/A01/X100"
              >
                Yes - view online
              </a>
            </td>
            <td
              className="govuk-table__cell"
            >
              Open
            </td>
            <td
              className="govuk-table__cell"
            >
              Yes
               (
              <a
                className="govuk-link"
                href="/organisations/A01/courses/X100/vacancies"
              >
                Edit
              </a>
              )
            </td>
          </tr>
        </tbody>
      </table>
      <h2
        className="govuk-heading-m"
      >
        <span
          className="govuk-caption-m"
        >
          Accredited body
        </span>
        Acme Accrediting
      </h2>
      <table
        className="govuk-table"
      >
        <thead
          className="jsx-621630794 govuk-table__head"
        >
          <tr
            className="jsx-621630794 govuk-table__row"
          >
            <th
              className="jsx-621630794 govuk-table__header w-33"
            >
              Course
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              UCAS Status
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Content
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Is it on 
              <abbr
                className="jsx-621630794"
                title="Find postgraduate teacher training"
              >
                Find
              </abbr>
              ?
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Applications
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Vacancies
            </th>
          </tr>
        </thead>
        <tbody
          className="govuk-table__body"
        >
          <tr
            className="govuk-table__row"
          >
            <td
              className="govuk-table__cell"
            >
              <a
                className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
              >
                Mathematics
                 (
                X101
                )
              </a>
              <span
                className="govuk-body-s"
              >
                QTS full time
              </span>
            </td>
            <td
              className="govuk-table__cell"
            >
              Not running
            </td>
            <td
              className="govuk-table__cell"
            />
            <td
              className="govuk-table__cell"
            >
              No
            </td>
            <td
              className="govuk-table__cell"
            >
              
            </td>
            <td
              className="govuk-table__cell"
            >
              
            </td>
          </tr>
        </tbody>
      </table>
      <h2
        className="govuk-heading-m"
      >
        <span
          className="govuk-caption-m"
        >
          Accredited body
        </span>
        Acme Other
      </h2>
      <table
        className="govuk-table"
      >
        <thead
          className="jsx-621630794 govuk-table__head"
        >
          <tr
            className="jsx-621630794 govuk-table__row"
          >
            <th
              className="jsx-621630794 govuk-table__header w-33"
            >
              Course
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              UCAS Status
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Content
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Is it on 
              <abbr
                className="jsx-621630794"
                title="Find postgraduate teacher training"
              >
                Find
              </abbr>
              ?
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Applications
            </th>
            <th
              className="jsx-621630794 govuk-table__header"
            >
              Vacancies
            </th>
          </tr>
        </thead>
        <tbody
          className="govuk-table__body"
        >
          <tr
            className="govuk-table__row"
          >
            <td
              className="govuk-table__cell"
            >
              <a
                className="govuk-link govuk-heading-s govuk-!-margin-bottom-0"
              >
                Physics
                 (
                X102
                )
              </a>
              <span
                className="govuk-body-s"
              >
                QTS full time
              </span>
            </td>
            <td
              className="govuk-table__cell"
            >
              New – not yet running
            </td>
            <td
              className="govuk-table__cell"
            />
            <td
              className="govuk-table__cell"
            >
              No
            </td>
            <td
              className="govuk-table__cell"
            >
              
            </td>
            <td
              className="govuk-table__cell"
            >
              
            </td>
          </tr>
        </tbody>
      </table>
    </section>
    <a
      className="govuk-button govuk-!-margin-bottom-2"
      href="https://docs.google.com/forms/d/e/1FAIpQLSeatJO2ZuqM-fnRJxEo6IIF0hZIU63JGnx0sDXO6Ulax7U_bA/viewform?usp=pp_url&entry.1033530353=tim.abell%2B4%40digital.education.gov.uk&entry.158771972=2C4"
      rel="noopener noreferrer"
      target="_blank"
    >
      Add a new course
    </a>
    <p
      className="govuk-body-s "
    >
      You’ll be taken to a Google Form to fill in your course details.
    </p>
  </main>,
]
`;
